### DISCUSSION TITLE
Computing number of inverted triangles in a mesh

### DISCUSSION THREAD

**Discussion by drathin4** [libigl/libigl#1993] [Q&A]
I want to compute the number of inverted triangles in a mesh. Is `flipped_triangles()` the right API for doing that ? Assuming that it is, the code that I wrote for computing inverted triangles is as follows:

1. igl::remove_unreferenced()
The reason for using above API is mentioned in `lscm.h` file. 
2. igl::lscm()
Used the code for example `502 (least square conformal mapping)` which gives 2D vertex coordinates `(uv map)` from 3D vertex coordinates : [https://github.com/libigl/libigl/blob/main/tutorial/502_LSCMParam/main.cpp](url)
3. igl::flipped_triangles() 
Eigen::VectorXi flippedTriangles.rows() gives the number.

With this procedure, I'm getting humongous number of inverted triangles. Typically such flipped/inverted triangles appear dark when visualized and I can't see so many dark areas in the mesh, hence it's difficult to believe that number. Am I using the correct API or the given APIs in the right sequence or I need to do something different ?

**Comment by drathin4** [comment]
In the interim I came up with 2 different algorithms to compute number of inverted triangles. But each are giving different answers. I want to know if any of them is right.

1. Signed areas:
igl::lscm() gives the same `uv map` described in the issue
igl::doublearea() -> gives signed areas
The number of triangles with -ve areas have CW orientation of edges, hence inverted.

2. Vector dot products:
igl::centroid() -> mesh centroid
igl::copyleft::cgal::points_inside_component() gives position of centroid wrt mesh `(inside/outside)`
Compute facet centroids individually using triangle centroid formula
Draw vector from each facet centroid to mesh centroid -> `v1`
igl::per_face_normals() -> `v2`
`
for (all facets)
    if (inside && v1 . v2 > 0)
            ++inverted_triangle
    else if (ouside && v1 . v2 < 0)
            ++inverted_triangle
`

**Reply by alecjacobson** [reply]
The difference between these two methods makes me question whether I understood what you're trying to compute. Is your mesh in 2D or 3D?

**Reply by drathin4** [reply]
It is a 3D surface (triangular) mesh

**Reply by alecjacobson** [reply]
Thanks. So, then, what do you mean by "inverted triangle"?

**Reply by drathin4** [reply]
By that I mean triangles whose normals are flipped (orientation is reversed). Such triangles appear `black` when visualized in MeshLab.

**Reply by alecjacobson** [reply]
ah hah, yes, this is different from what I understood. You might look at the bfs_orient method that works for many models. You can also try reorient_facets_raycast

**Reply by drathin4** [reply]
Thanks for your help. `bfs_orient()` gave better results than `reorient_facets_raycast()`. The latter was also much slower. Just out of curiosity, what does the API `flipped_triangles()` do ? I saw in the implementation that it computes the sign of the determinant. But I don't think it checks for triangles with inverted normals.
